### PR TITLE
chore: purity in every `exp`

### PR DESCRIPTION
**Pull request by ChrBondi**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
@mlutze Could you do a quick pass just to see if anything is amiss? Then I will do a pass after.

**Comment by magnus-madsen** [comment]
@ChrBondi You have some failing tests.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 1]
```diff
@@ -36,168 +36,206 @@ object SimplifiedAst {
...
+      def purity: Purity = Pure
     }
 
-    case class Var(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Var(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression {
+      def purity: Purity = Pure
+    }
 
-    case class Def(sym: Symbol.DefnSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Def(sym: Symbol.DefnSym, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:114]: Are there cases when this can be impure?

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 2]
```diff
@@ -36,168 +36,206 @@ object SimplifiedAst {
...
 
-    case class ArrayStore(base: SimplifiedAst.Expression, index: SimplifiedAst.Expression, elm: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class ArrayStore(base: SimplifiedAst.Expression, index: SimplifiedAst.Expression, elm: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression {
+      def purity: Purity = Impure
+    }
 
-    case class ArrayLength(base: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class ArrayLength(base: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression {
+      def purity: Purity = Impure
+    }
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:190]: Array length is actually pure (at least in the front end) since the length of an array never changes.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 3]
```diff
@@ -36,168 +36,206 @@ object SimplifiedAst {
...
 
-    case class Ref(exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Ref(exp: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class Deref(exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Deref(exp: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class Assign(exp1: SimplifiedAst.Expression, exp2: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Assign(exp1: SimplifiedAst.Expression, exp2: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:200]: These are always Impure, right?

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 4]
```diff
@@ -36,168 +36,206 @@ object SimplifiedAst {
...
+    case class InvokeStaticMethod(method: Method, args: List[SimplifiedAst.Expression], tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class GetField(field: Field, exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class GetField(field: Field, exp: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class PutField(field: Field, exp1: SimplifiedAst.Expression, exp2: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class PutField(field: Field, exp1: SimplifiedAst.Expression, exp2: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class GetStaticField(field: Field, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class GetStaticField(field: Field, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:216]: These JVM methods should likely all be Impure.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 5]
```diff
@@ -36,168 +36,206 @@ object SimplifiedAst {
...
+    case class GetChannel(exp: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class PutChannel(exp1: SimplifiedAst.Expression, exp2: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class PutChannel(exp1: SimplifiedAst.Expression, exp2: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class SelectChannel(rules: List[SimplifiedAst.SelectChannelRule], default: Option[SimplifiedAst.Expression], tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class SelectChannel(rules: List[SimplifiedAst.SelectChannelRule], default: Option[SimplifiedAst.Expression], tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
-    case class Spawn(exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Spawn(exp: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:228]: Channel functions are also probably all always impure.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 6]
```diff
@@ -36,168 +36,206 @@ object SimplifiedAst {
...
 
-    case class Lazy(exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Lazy(exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression {
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:230]: Lazy expressions are always pure in the front-end. But maybe their simplification is always impure?

**Review by mlutze — COMMENTED** [review]
Some things to consider about the purity of the expressions, plus a style hint.

Since I'm unfamiliar with both the project and this part of the compiler, I wasn't too sure about the more complex stuff in the Simplifier.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 7]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
+      def purity: Purity = Pure
     }
 
-    case class Var(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Var(sym: Symbol.VarSym, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression {
+      def purity: Purity = Pure
+    }
 
-    case class Closure(sym: Symbol.DefnSym, freeVars: List[FreeVar], tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Closure(sym: Symbol.DefnSym, freeVars: List[FreeVar], tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:113]: A closure is always pure.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 8]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
 
-    case class Binary(sop: SemanticOperator, op: BinaryOperator, exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Binary(sop: SemanticOperator, op: BinaryOperator, exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
 
-    case class IfThenElse(exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, exp3: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class IfThenElse(exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, exp3: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
 
-    case class Branch(exp: Expression, branches: Map[Symbol.LabelSym, LiftedAst.Expression], tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Branch(exp: Expression, branches: Map[Symbol.LabelSym, LiftedAst.Expression], tpe: Type, loc: SourceLocation) extends LiftedAst.Expression  {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:131]: Branching can be pure.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 9]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
+      def purity: Purity = Impure
+    }
 
-    case class JumpTo(sym: Symbol.LabelSym, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class JumpTo(sym: Symbol.LabelSym, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:135]: The same for JumpTo, can also be pure.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 10]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
+      def purity: Purity = Impure
+    }
 
     case class Let(sym: Symbol.VarSym, exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:139]: Now of course purity should refer to the purity of the whole expression, not of exp1.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 11]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
+      def purity: Purity = Impure
+    }
 
-    case class ArrayStore(base: LiftedAst.Expression, index: LiftedAst.Expression, elm: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class ArrayStore(base: LiftedAst.Expression, index: LiftedAst.Expression, elm: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression {
+      def purity: Purity = Impure
+    }
 
-    case class ArrayLength(base: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class ArrayLength(base: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:181]: Whether ArrayLength is pure depends on whether base is pure, in other words, you need the field.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 12]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
+    }
 
-    case class Ref(exp: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Ref(exp: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
 
-    case class Deref(exp: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Deref(exp: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
 
-    case class Assign(exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Assign(exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:193]: Ref, Deref, and Assign are all Impure.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 13]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
+    case class PutChannel(exp1: LiftedAst.Expression, exp2: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
 
-    case class SelectChannel(rules: List[LiftedAst.SelectChannelRule], default: Option[LiftedAst.Expression], tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class SelectChannel(rules: List[LiftedAst.SelectChannelRule], default: Option[LiftedAst.Expression], tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
 
-    case class Spawn(exp: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Spawn(exp: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
 
-    case class Lazy(exp: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Lazy(exp: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression  {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:223]: Lazy is always pure.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala [hunk 14]
```diff
@@ -35,162 +35,200 @@ object LiftedAst {
...
+      def purity: Purity = Impure
+    }
 
-    case class Force(exp: LiftedAst.Expression, tpe: Type, loc: SourceLocation) extends LiftedAst.Expression
+    case class Force(exp: LiftedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends LiftedAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:227]: And force is pure.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala (hunk 1 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:114]: This is always pure.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:190]: It depends on whether base is pure. Thus it needs a field.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala (hunk 3 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:200]: Yes

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala (hunk 4 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:216]: For JVM, it depends. E.g. String.toUpperCase is pure.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala (hunk 5 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:228]: Yes

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala (hunk 6 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:230]: Lazy/force are pure.

**Review by magnus-madsen — COMMENTED** [review]
Added first batch of comments.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala (hunk 8 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:131]: What I mean is that it depends on the other expressions, i.e. it needs a field.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala (hunk 9 above)
Comment (ChrBondi) [reply main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala:135]: What exp would you base it?

**Comment by magnus-madsen** [comment]
Lets get this merged in next, since its big :)

I see two conflicts, if you resolve those, I can have another look.

File: main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala (hunk 9 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala]: Ah- thats a problem. Let me think...

Comment (ChrBondi) [reply main/src/ca/uwaterloo/flix/language/ast/LiftedAst.scala]: I added something that might be right if you could have a look at it

File: main/src/ca/uwaterloo/flix/language/ast/Purity.scala [hunk 15]
```diff
@@ -32,4 +32,15 @@ object Purity {
    * Represents an impure expression (i.e. an expression that could potentially have side-effects).
    */
   case object Impure extends Purity
+
+  /**
+   * Infix operator `combine` merges purities `self` and `that`
+   * A merged purity is only pure if both `self` and `that` are pure, otherwise it is always impure.
+   */
+  implicit class compare(self: Purity) extends Purity {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Purity.scala:40]: We never use implicit classes. They are really a misfeature.

Where is this used and can it not be a normal function?

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Purity.scala:40]: This was a result of my suggestion. I thought an infix would be a clean syntax for combining multiple purities. Though I meant to add it as a method in the Purity trait :)

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Purity.scala:40]: We can add it to the purity trait, but I prefer just a good ol' regular function somewhere.

File: main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala [hunk 16]
```diff
@@ -35,162 +36,214 @@ object OccurrenceAst {
...
-    case class Force(exp: OccurrenceAst.Expression, tpe: Type, loc: SourceLocation) extends OccurrenceAst.Expression
+    case class Force(exp: OccurrenceAst.Expression, tpe: Type, loc: SourceLocation) extends OccurrenceAst.Expression {
+      def purity: Purity = Pure
+    }
 
-    case class HoleError(sym: Symbol.HoleSym, tpe: Type, loc: SourceLocation) extends OccurrenceAst.Expression
+    case class HoleError(sym: Symbol.HoleSym, tpe: Type, purity: Purity, loc: SourceLocation) extends OccurrenceAst.Expression
 
-    case class MatchError(tpe: Type, loc: SourceLocation) extends OccurrenceAst.Expression
+    case class MatchError(tpe: Type, purity: Purity, loc: SourceLocation) extends OccurrenceAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/OccurrenceAst.scala:246]: I think we should always treat HoleError and MatchError as impure.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 17]
```diff
@@ -36,168 +36,222 @@ object SimplifiedAst {
...
+      def purity: Purity = Pure
+    }
 
-    case class Def(sym: Symbol.DefnSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Def(sym: Symbol.DefnSym, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression {
+      def purity: Purity = Pure
+    }
 
-    case class Lambda(fparams: List[SimplifiedAst.FormalParam], exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Lambda(fparams: List[SimplifiedAst.FormalParam], exp: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:118]: A lambda is always pure.

File: main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala [hunk 18]
```diff
@@ -36,168 +36,222 @@ object SimplifiedAst {
...
 
-    case class Apply(exp: SimplifiedAst.Expression, args: List[SimplifiedAst.Expression], tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class Apply(exp: SimplifiedAst.Expression, args: List[SimplifiedAst.Expression], tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
 
     @IntroducedBy(ClosureConv.getClass)
-    case class LambdaClosure(fparams: List[SimplifiedAst.FormalParam], freeVars: List[FreeVar], exp: SimplifiedAst.Expression, tpe: Type, loc: SourceLocation) extends SimplifiedAst.Expression
+    case class LambdaClosure(fparams: List[SimplifiedAst.FormalParam], freeVars: List[FreeVar], exp: SimplifiedAst.Expression, tpe: Type, purity: Purity, loc: SourceLocation) extends SimplifiedAst.Expression
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/SimplifiedAst.scala:123]: A lambda closure is always pure.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 19]
```diff
@@ -50,285 +50,289 @@ object Simplifier {
...
-
-      case TypedAst.Expression.Int32(lit, loc) => SimplifiedAst.Expression.Int32(lit, loc)
-
-      case TypedAst.Expression.Int64(lit, loc) => SimplifiedAst.Expression.Int64(lit, loc)
-
-      case TypedAst.Expression.BigInt(lit, loc) => SimplifiedAst.Expression.BigInt(lit, loc)
+     * Translates the given expression `exp0` to the SimplifiedAst.
+     */
+    def visitExp(exp0: TypedAst.Expression): SimplifiedAst.Expression = {
+      val purity = isPure(exp0.eff)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:56]: Please don't do this. Its bad form, just do it locally.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:56]: Because visitExp is so long, if you do this purity is essentially a global variable, which we dont want.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 20]
```diff
@@ -50,285 +50,289 @@ object Simplifier {
...
+      exp0 match {
+        case TypedAst.Expression.Var(sym, tpe, loc) => SimplifiedAst.Expression.Var(sym, tpe, loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:58]: I am confused by why you re-ordered everything in this file. This makes it hard to review.

Sometimes we do refactor, but then we do it in a separate PR.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:58]: OK, I can't really review it because there are too many irrelevant changes. I suggest you redo the implementation of this phase, keeping the delta minimal.

**Review by magnus-madsen — COMMENTED** [review]
OK, You are almost done. One more iteration and then we can merge !:)

**Comment by magnus-madsen** [comment]
Looks good, I just need to do a pass over Simplifier and then one final iteration, and I think we can merge :)

**Comment by magnus-madsen** [comment]
Hopefully we can have that done by the end of today.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 21]
```diff
@@ -342,8 +340,15 @@ object Simplifier {
...
       case TypedAst.Pattern.BigInt(lit, loc) => SimplifiedAst.Expression.BigInt(lit, loc)
       case TypedAst.Pattern.Str(lit, loc) => SimplifiedAst.Expression.Str(lit, loc)
-      case TypedAst.Pattern.Tag(sym, tag, p, tpe, loc) => SimplifiedAst.Expression.Tag(sym, tag, pat2exp(p), tpe, loc)
-      case TypedAst.Pattern.Tuple(elms, tpe, loc) => SimplifiedAst.Expression.Tuple(elms map pat2exp, tpe, loc)
+      case TypedAst.Pattern.Tag(sym, tag, p, tpe, loc) =>
+        val e = pat2exp(p)
+        SimplifiedAst.Expression.Tag(sym, tag, e, tpe, e.purity, loc)
+      case TypedAst.Pattern.Tuple(elms, tpe, loc) =>
+        val es = elms.map(pat2exp)
+        val purity = es.foldLeft[Purity](Pure){
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:348]: Introduce a function, after `combine`, called `combineAll` which takes a list of Purities and combines them.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 22]
```diff
@@ -1041,10 +1068,19 @@ object Simplifier {
...
       Pure
     else
       Impure
   }
+
+  /**
+   * Merges purities `p1` and `p2`
+   * A merged purity is only pure if both `p1` and `that` are pure, otherwise it is always impure.
+   */
+  def combinePurities(p1: Purity, p2: Purity): Purity = (p1, p2) match {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:1082]: Rename to `combine`. 
Introduce `combine`  function that takes 3 purities (and use it).
Introduce other function `combineAll` that takes a list of purities.

File: main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala [hunk 23]
```diff
@@ -475,28 +482,36 @@ object Simplifier {
           // Success case: evaluate the match body.
           val success = visitExp(body)
 
+          val purity = combinePurities(effectToPurity(guard.eff), effectToPurity(body.eff))
+
           // Failure case: Jump to the next label.
-          val failure = SimplifiedAst.Expression.JumpTo(next, tpe, loc)
+          val failure = SimplifiedAst.Expression.JumpTo(next, tpe, purity, loc)
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/Simplifier.scala:488]: Intermediate solution (which is correct, but imprecise): Compute the purity of every match rule in `rules`

**Comment by magnus-madsen** [comment]
Merged. Yes!
